### PR TITLE
Porting SessionTokenMismatchRetryPolicy from .Net-SDK

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
In the .Net-SDK a retry policy had been added to ensure local retries in case of 404/1002 in the ConsistencyReader/Writer stack. The StoreReader itself is iterating through replicas already - so this retry policy is only applicable when initially no local replica satisfies the session token provided - in .Net the main driver for adding this retry policy has been that we saw errors for uncommitted writes resulting in request-level regional fail-over where the SessionTokenMismatchRetryPolicy (with pretty aggressive retry intervals) results overall in better experience for multi-regional accounts - and is a significant improvement of course for single-region accounts.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java [hunk 1]
```diff
@@ -72,6 +72,20 @@
     private static final int DEFAULT_QUERY_PLAN_RESPONSE_TIMEOUT_IN_SECONDS = 5;
     private static final int DEFAULT_ADDRESS_REFRESH_RESPONSE_TIMEOUT_IN_SECONDS = 5;
 
+    // SessionTokenMismatchRetryPolicy Constants
+    // TODO fabianm - do we really want to keep these config overrides like in .Net
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:76]: In .Net we allow overriding these defaults via Environment variables - this is the reason why I added it here as well. But to be honest I don't expect this to be an often used extensibility - most likely the configuration was allowed in .Net for Compute/Mongo/Cassandra - so I would be ok with removing and hard-coding the settings instead as well. Let me know what you think...

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionTokenMismatchRetryPolicy.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+            LOGGER.debug(
+                "SessionTokenMismatchRetryPolicy not retrying because StatusCode or SubStatusCode not found.");
+
+            return Mono.just(ShouldRetryResult.noRetry());
+        }
+
+        if (this.waitTimeTimeoutHelper.isElapsed()) {
+
+            LOGGER.info(
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionTokenMismatchRetryPolicy.java:61]: info vs warn?

**Review by kirankumarkolli — APPROVED** [review]

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java [hunk 3]
```diff
@@ -72,6 +72,20 @@
...
+    private static final String DEFAULT_SESSION_TOKEN_MISMATCH_WAIT_TIME_IN_MILLISECONDS_NAME =
+        "session-token-mismatch-policy-default-wait-time-milliseconds";
+    private static final int DEFAULT_SESSION_TOKEN_MISMATCH_WAIT_TIME_IN_MILLISECONDS = 5000;
+
+    private static final String DEFAULT_SESSION_TOKEN_MISMATCH_INITIAL_BACKOFF_TIME_IN_MILLISECONDS_NAME =
+        "session-token-mismatch-policy-default-initial-backoff-milliseconds";
+    private static final int DEFAULT_SESSION_TOKEN_MISMATCH_INITIAL_BACKOFF_TIME_IN_MILLISECONDS = 5;
+
+    private static final String DEFAULT_SESSION_TOKEN_MISMATCH_MAXIMUM_BACKOFF_TIME_IN_MILLISECONDS_NAME =
+        "session-token-mismatch-policy-default-maximum-backoff-milliseconds";
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:86]: the new config names not consistent with the existing ones.
```suggestion
        "COSMOS.SESSION_TOKEN_MISMATCH_POLICY_DEFAULT_MAX_BACKOF_IN_MILLISECONDS"
```

same for the other configs

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionTokenMismatchRetryPolicy.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                this.retryCount);
+
+            return Mono.just(ShouldRetryResult.noRetry());
+        }
+
+        Duration effectiveBackoff = Duration.ZERO;
+
+        // Don't penalize first retry with delay
+        if (this.retryCount.getAndIncrement() > 0)
+        {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionTokenMismatchRetryPolicy.java:73]: code style: in java "{" should be on the same line as if.

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyReaderTest.java [hunk 5]
```diff
@@ -424,6 +423,122 @@ public void sessionNotAvailableFromSomeReplicasThrowingNotFound_FindReplicaSatis
...
         validateSuccess(storeResponseSingle, validator);
+        transportClientWrapper.verifyNumberOfInvocations(4);
+    }
+
+    /**
+     * reading in session consistency, if the requested session token cannot be supported by some replicas
+     * tries others till we find a replica which can support the given session token
+     */
+    @Test(groups = "unit")
+    public void sessionNotAvailableFromAllReplicasThrowingNotFound_FindReplicaSatisfyingRequestedSessionOnRetry() {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyReaderTest.java:434]: great test. thanks.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyReader.java [hunk 6]
```diff
@@ -235,7 +231,9 @@ public ConsistencyReader(
 
             case Any:
                 if (targetConsistencyLevel.v == ConsistencyLevel.SESSION) {
-                    return this.readSessionAsync(entity, desiredReadMode);
+                    return BackoffRetryUtility.executeRetry(
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyReader.java:234]: we should consider doing a quick perf test for session consistency to see the perf impact of having retry logic hooked up here?

**Review by moderakh — APPROVED** [review]
LGTM.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionTokenMismatchRetryPolicy.java (hunk 4 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionTokenMismatchRetryPolicy.java:73]: ACK - Fixed in next iteartion

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyReader.java (hunk 6 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyReader.java:234]: Will do that before merging

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java (hunk 3 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:86]: Fixed in next iteration

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionTokenMismatchRetryPolicy.java (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/SessionTokenMismatchRetryPolicy.java:61]: ACK - Fixed in next iteration

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyReader.java (hunk 6 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyReader.java:234]: **_Results - are in the same ballpark - if any regression it is minimal (around 1%)....._**

**Baseline**
=============

_[2000000] operations performed in [296] seconds._

-- Timers ----------------------------------------------------------------------
latency
             count = 2000000
     1-minute rate = 6770.75 calls/second
     5-minute rate = 4997.96 calls/second
              95% <= 1.98 milliseconds
              98% <= 2.75 milliseconds
              99% <= 4.39 milliseconds
            99.9% <= 11.47 milliseconds


**Candidate (with SessionTokenMismatchRetryPolicy)**
==================================================

_[2000000] operations performed in [301] seconds._

-- Timers ----------------------------------------------------------------------
latency
             count = 2000000
     1-minute rate = 6691.68 calls/second
     5-minute rate = 4945.42 calls/second
              95% <= 2.09 milliseconds
              98% <= 3.16 milliseconds
              99% <= 4.52 milliseconds
            99.9% <= 11.01 milliseconds

**Pull request merged**